### PR TITLE
use -Ywarn-unused for all Scala code

### PR DESCRIPTION
**Pull request by S11001001**
Typically, when some code is unused, it's simply that we didn't notice, or carry through the follow-up cleanup from earlier changes. There are also cases where we believe that something is used, when really it is not, for which there are two reasonable reactions: we should use it, or we should no longer have code that makes it look like it is used.

Here we enable `-Ywarn-unused`, which was in `lf_scalacopts`, in all Scala code, and clean up or suppress warnings as appropriate.

Some removal split off to separate PRs:

* [x] merge #6980
* [x] merge #6981
* [x] merge #6982
* [x] merge #6983
* [x] merge #6985
* [x] merge #6987
* [x] merge #6992
* [x] merge #7031
* [x] merge #7032
* [x] merge #7035
* [x] merge #7051

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/CommandClientIT.scala [hunk 1]
```diff
@@ -127,7 +127,7 @@ final class CommandClientIT
   private def readExpectedElements[T](
       src: Source[T, NotUsed],
       expected: Set[T],
-      timeLimit: Span = 3.seconds): Future[(Set[T], Set[T])] =
```
Comment (cocreature) [inline ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/CommandClientIT.scala:130]: Why did you remove the default? Seems like a reasonable thing to keep.

File: bazel_tools/scala.bzl [hunk 2]
```diff
@@ -71,9 +71,7 @@ common_scalacopts = [
     # Gives a warning for functions declared as returning Unit, but the body returns a value
     "-Ywarn-value-discard",
     "-Ywarn-unused-import",
-    # unfortunately give false warning for the `(a, b) = someTuple`
```
[the remark this answers is not in the document]
Comment (S11001001) [reply bazel_tools/scala.bzl:75]: This `for` issue still exists, though more is required to encounter it than simply making a binding. I have included in this PR `silent` annotations for all `1` cases in the codebase where we trigger this issue.

I think these sorts of things are a big problem if you don't have `silent`, and become significantly less so when you do.

File: ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/CommandClientIT.scala (hunk 1 above)
Comment (S11001001) [reply ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/CommandClientIT.scala:130]: It is unused. In that it is like a private definition alongside that is never used, hence `-Ywarn-unused` including checks for these sorts of things.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala [hunk 3]
```diff
@@ -251,7 +251,11 @@ final case class ScenarioRunner(
             missingWith(SErrorCrash(s"contract $acoid not effective, but we found its key!"))
           case ScenarioLedger.LookupContractNotActive(_, _, _) =>
             missingWith(SErrorCrash(s"contract $acoid not active, but we found its key!"))
-          case ScenarioLedger.LookupContractNotVisible(coid, tid, observers @ _, stakeholders) =>
+          case ScenarioLedger.LookupContractNotVisible(
+              coid,
+              tid @ _,
+              observers @ _,
+              stakeholders) =>
```
Comment (remyhaemmerle-da) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:258]:
```suggestion
              stakeholders,
            ) =>
```

File: compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Context.scala [hunk 4]
```diff
@@ -197,7 +197,7 @@ class Context(val contextId: Context.ContextId) {
     val compiledPackages =
       PureCompiledPackages(allPackages, defns, Compiler.FullStackTrace, Compiler.NoProfile)
     val (scriptPackageId, _) = allPackages.find {
```
Comment (remyhaemmerle-da) [inline compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Context.scala:199]: collect ?

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala (hunk 3 above)
Comment (S11001001) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala]: ead2923dd0

File: compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Context.scala (hunk 4 above)
Comment (S11001001) [reply compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Context.scala]: Why not. ead2923dd0

**Comment by hurryabit** [comment]
I'm big fan of such warnings. Particularly in codebases where amateurs like me run around and change things they don't have the deepest understanding of. Thanks a lot.

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala [hunk 5]
```diff
@@ -443,7 +443,7 @@ object Repl {
 
     parser.parseExprs[this.type](args.mkString(" ")) match {
 
-      case Left(error) =>
+      case Left(error @ _) =>
```
Comment (SamirTalwar-DA) [inline daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala:446]: What does this `x @ _` syntax do? I would have thought it would just be the same as `x`.

Unfortunately, it's very hard to search the web for this kind of syntax, so I have to ask.

File: bazel_tools/scala.bzl [hunk 6]
```diff
@@ -135,7 +133,6 @@ plugin_scalacopts = [
 # []
 # ^ means nothing to remove
 lf_scalacopts = [
-    "-Ywarn-unused",
 ]
```
Comment (SamirTalwar-DA) [inline bazel_tools/scala.bzl:136]: Are we keeping `lf_scalacopts` around for a reason?

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala (hunk 5 above)
Comment (leo-da) [reply daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala:446]: it is the same as `_`... but someone may say it is more readable because it includes a descriptive var name (unused).

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala [hunk 7]
```diff
@@ -115,6 +116,7 @@ class SubmissionValidator[LogResult] private[validator] (
       )
     }
 
+  @silent(" ignored .* is never used") // matches runValidation signature
   private def commit(
       logEntryId: DamlLogEntryId,
       ignored: Any,
```
Comment (leo-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/SubmissionValidator.scala:122]: would not `_: Any`, do the same thing as the above `@silent` annotation?

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala (hunk 5 above)
Comment (S11001001) [reply daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala:446]: Sometimes it is helpful, and sometimes it is just noise; I have used one or the other based on my judgment in each case.

File: bazel_tools/scala.bzl (hunk 6 above)
Comment (S11001001) [reply bazel_tools/scala.bzl:136]: It turned out to be very useful, since it made it possible to fully migrate part of the codebase to the flag (in fact, I wish there were more partitions of the codebase with cross-project scalacopts sets like this one), it would be a hassle to remove, and it would be a hassle to restore.

Removing it would be like saying about a Scala build that supported cross-version builds "well, we support the latest Scala version, and it's not like they're ever going to release a new Scala version, right?"

**Comment by stefanobaghino-da** [comment]
I have no idea why we haven't had this so far, thanks!
